src/rsa/blinding.rs

```diff
@@ -12,52 +12,127 @@
+                    where F: FnOnce(bigint::ElemDecoded<N>)
+                                    -> Result<bigint::ElemDecoded<N>,
+                                              error::Unspecified> {
+        let old_contents = core::mem::replace(&mut self.0, None);
+
+        let new_contents = try!(match old_contents {
+            Some(Contents {
+                blinding_factor,
+                blinding_factor_inv,
+                remaining,
+            }) => {
+                if remaining > 0 {
+                    // Update the existing blinding factor by squaring it, as
+                    // suggested in the paper.
+                    let blinding_factor =
+                        try!(bigint::elem_squared(blinding_factor, n));
+                    let blinding_factor_inv =
+                        try!(bigint::elem_squared(blinding_factor_inv, n));
+                    Ok(Contents {
+                        blinding_factor: blinding_factor,
+                        blinding_factor_inv: blinding_factor_inv,
+                        remaining: remaining - 1,
+                    })
+                } else {
+                    // Create a new, independent blinding factor.
+                    reset(blinding_factor, blinding_factor_inv, e, n, rng)
+                }
+            },
+
+            None => {
+                let elem1 = try!(bigint::Elem::zero());
+                let elem2 = try!(bigint::Elem::zero());
+                reset(elem1, elem2, e, n, rng)
+            },
+        });
+
+        // Blind `x`.
+        let x =
+            try!(bigint::elem_mul_mixed(&new_contents.blinding_factor, x, n));
+
+        let x = try!(f(x));
+
+        // Unblind `x`.
+        let x =
+            try!(bigint::elem_mul_mixed(&new_contents.blinding_factor_inv, x,
+                                        n));
+
+        let _ = core::mem::replace(&mut self.0, Some(new_contents));
+
+        Ok(x)
     }
 
     #[cfg(test)]
-    pub fn counter(&self) -> u32 { unsafe { (*self.0).counter } }
-
-    pub fn as_mut_ref(&mut self) -> &mut BN_BLINDING { unsafe { &mut *self.0 } }
+    pub fn remaining(&self) -> usize {
+        match &self.0 {
+            &Some(Contents { remaining, .. }) => remaining,
+            &None => { 0 },
+        }
+    }
 }
 
-/// Needs to be kept in sync with `bn_blinding_st` in `crypto/rsa/blinding.c`.
-#[allow(non_camel_case_types)]
-#[repr(C)]
-pub struct BN_BLINDING {
-    a: *mut bigint::BIGNUM,
-    ai: *mut bigint::BIGNUM,
-    counter: u32,
-}
+fn reset(arbitrary1: bigint::Elem<N>, arbitrary2: bigint::Elem<N>,
+         e: &bigint::OddPositive, n: &bigint::Modulus<N>,
+         rng: &rand::SecureRandom) -> Result<Contents, error::Unspecified> {
+    // Use `into_elem_decoded_montgomery_encoded` to grab the underling
+    // `BIGNUM` to avoid a superfluous `malloc()` & `free()`.
+    let mut random = arbitrary1.into_elem_decoded_montgomery_encoded();
```

Comment:
Perhaps `Storage` might be a better name than `Arbitrary`, since really it's entire purpose is to allow the `BIGNUM` to be reused.

Then we could have `elem_random<F: Field>(storage: Storage<F>, m: Modulus<F>) -> ElemDecoded<F>` and `elem_inverse<F>(storage: Storage<F>, a: &ElemDecoded<F>, m: &Modulus<F>) -> ElemDecoded<F>`. As a bonus, those functions would become pure functions,.

In contrast, we could use the name `Arbitrary` but it might be misleading to pass an "arbitrary" value to these functions. Especially the random generation taking an `Arbitrary` kind of implies that values is used for (re)seeding or something.

WDYT?